app/workers/user_cleaner.rb

```diff
@@ -0,0 +1,136 @@
+# Deletes inactive users from the database.
+# See [1] for a description of how the flow works on a high level.
+#
+# Users have a deletion_date field that is nil by default. It is set to a future
+# date if the user has been inactive for too long (i.e. hasn't logged in).
+# Before the deletion date is reached, we send warning mails. If users log in
+# before the deletion date, that date is reset to nil such that the user is not
+# deleted. If the user is still inactive on the deletion date, the user is
+# ultimately deleted.
+#
+# [1] https://github.com/MaMpf-HD/mampf/issues/410#issuecomment-2136875776
+class UserCleaner
+  # The maximum number of users that can be deleted in one run.
+  # This is equivalent to the maximum of number of deletion dates set in one run.
+  #
+  # This flag can be used to prevent too many mails from being sent at once.
+  # Keep in mind that the mail server also handles other mails, e.g. notification
+  # mails etc., so we might want to set the limit very low here such that our
+  # mail server is not marked as "spam server".
+  #
+  # Note that this is just a soft limit, i.e. the actual number of deletion
+  # warning mails sent on a given day might be higher than this number:
+  # - If on a given day the cronjob is not run (for whatever reason),
+  #   we have more users with a deletion date lying in the past than
+  #   MAX_DELETIONS_PER_RUN. However, we don't send an additional mail once
+  #   the user is deleted, so this shouldn't be a problem.
+  # - Despite that there cannot be more than MAX_DELETIONS_PER_RUN users with the
+  #   same deletion date, warning mails might be sent on the same date to users
+  #   with varying deletion dates, since the 40-, 14-, 7- and 2-day warning mails
+  #   can overlap temporally.
+  MAX_DELETIONS_PER_RUN = ENV.fetch("MAX_DELETIONS_PER_RUN").to_i
+
+  # The threshold for inactive users. Users who have not logged in for this time
+  # are considered inactive.
+  INACTIVE_USER_THRESHOLD = 6.months
+
+  # Returns all users who have been inactive for INACTIVE_USER_THRESHOLD months,
+  # i.e. their last sign-in date is more than INACTIVE_USER_THRESHOLD months ago.
+  def inactive_users
+    User.where("last_sign_in_at < ?", INACTIVE_USER_THRESHOLD.ago)
```

Comment:
If you have users that never logged in since #553 then their `last_sign_in_at` column will bei `nil` which means that they will not be counted as inactive here.

Reply:
Wow, thanks for noting that, that's really important! Fixed in 25041698ef